src/frontend/src/storage/index.ts

```diff
@@ -533,18 +547,97 @@ const StorageV3 = z.object({
   hasher: z.instanceof(CryptoKey),
 });
 
-const readIndexedDBV3 = async (): Promise<Storage | undefined> => {
-  const item: unknown = await idbGet(IDB_KEY_V3);
+const readIndexedDBV3 = (): Promise<StorageV3 | undefined> => {
+  return readGeneric({ idbKey: IDB_KEY_V3, storageType: StorageV3 });
+};
+
+const writeIndexedDBV3 = async (storage: StorageV3) => {
+  await idbSet(IDB_KEY_V3, storage);
+};
+
+const migratedV3 = async (): Promise<Storage | undefined> => {
+  const readStorage = await readIndexedDBV3();
+
+  if (isNullish(readStorage)) {
+    return undefined;
+  }
+
+  // No known principals
+  const migratedAnchors: Anchors = {};
+
+  for (const userNumber in readStorage.anchors) {
+    const oldAnchor = readStorage.anchors[userNumber];
+    migratedAnchors[userNumber] = { ...oldAnchor, knownPrincipals: [] };
+  }
+
+  return { anchors: migratedAnchors, hasher: readStorage.hasher };
+};
+
+const toV3 = (storage: Storage): StorageV3 => {
+  const migratedAnchors: AnchorsV3 = {};
+  for (const userNumber in storage.anchors) {
+    const oldAnchor = storage.anchors[userNumber];
+    migratedAnchors[userNumber] = { ...oldAnchor, knownPrincipals: [] };
+  }
+
+  return { anchors: migratedAnchors, hasher: storage.hasher };
+};
+
+/**
+ * V4, indexeddb["ii-storage-v4"] = { anchors: { 20000: ... } }
+ * */
+type StorageV4 = z.infer<typeof StorageV4>;
```

Comment:
@frederikrothenberger Why is there a V3 and a V4?

Reply:
The requirements have changed. The V3 (which was in use since December 2023) no longer stored enough information for the auto-selection for the most recently used principal.

Hence it was changed to V4, which stores the principal and origin digests separately. We still maintain compatibility with old storage layouts to prevent people losing their prepopulated identity numbers.